### ISSUE TITLE
Symfony: Class 'Cpdf' not found

### ISSUE THREAD

**Issue by ThomasLandauer** [dompdf/dompdf#1132]
I installed dompdf with `composer require dompdf/dompdf:0.7.x@beta`, as explained here: https://github.com/dompdf/dompdf/blob/master/README.md

However, Symfony 2.8.3 gives me this error:
`Fatal error: Class 'Cpdf' not found`

I can fix this by adding
`require_once __DIR__ . '/../../../vendor/dompdf/dompdf/lib/Cpdf.php';`
But this looks like a workaround....

**Comment by bsweeney** [comment]
I've not tried with Symfony specifically. Presumably you're using the composer autoloader?

**Comment by ThomasLandauer** [comment]
I guess, yes. At least, I've installed other libraries using composer and they worked out of the box. And `vendor/autoload.php` is existing and was modified 2 hours ago.
And dompdf got added to my `composer.json`:
````
"require": {
    ...
    "dompdf/dompdf": "0.7.x@beta"
    },

**Comment by bsweeney** [comment]
I'll have to set up a test environment to see what's going on.

**Comment by bsweeney** [comment]
How are you using dompdf within Symfony?

**Comment by ThomasLandauer** [comment]
For now, just straightforward in a Controller:
````php
$dompdf = new Dompdf();
$dompdf->loadHtml($this->renderView('pdf_test.html.twig')); // render() causes some HTTP Cache header to be shown
$dompdf->render();
$response = new Response ($dompdf->stream('dompdf.pdf'));
return $response;

**Comment by themao** [comment]
@ThomasLandauer I would suggest using symfony's bundles for this. For example this one https://github.com/SlikNL/DompdfBundle

**Comment by themao** [comment]
@bsweeney funny stuff, I'm observing the same exception, too. But not with symfony - I'm using Phalcon. Nevertheless, I was installing the library via Composer. I suppose there is some issue with it

**Comment by themao** [comment]
So it appears that this was solved already a year ago https://github.com/dompdf/dompdf/issues/919, but in 0.7.x@beta it definitely does not work. Googling around a little bit said the problem is that composer does not include files without namespaces (according to PSR-4) so I had to include it manually:

`"autoload": {
        "classmap": ["vendor/dompdf/dompdf/lib"]
    }`

**Comment by bsweeney** [comment]
Wow, thanks for the research @themao! @beta is definitely the wrong tag. I don't know why that slipped under the radar for so long :(

**Comment by ThomasLandauer** [comment]
Ahem, when I do `composer require dompdf/dompdf:0.7.x@beta3` - as explained at https://github.com/dompdf/dompdf/blob/master/README.md - I get:
`Could not parse version constraint 0.7.x@beta3: Invalid version string "0.7.x@beta3"`

**Comment by bsweeney** [comment]
@ThomasLandauer ... d'oh!

OK, I threw that in there quickly but obviously it's wrong. Can you try using just `v0.7.0-beta3`. I need to refresh myself on what works best with composer.

**Comment by ThomasLandauer** [comment]
Yeah, thanks, this now worked: `composer require dompdf/dompdf:v0.7.0-beta3`
Well, at least regarding the composer part ;-) I'll check and report on the other issues.

**Comment by bsweeney** [comment]
thanks, I'll get the docs updated today with an *actual* correct version tag

**Comment by ThomasLandauer** [comment]
Summarizing: With the new composer command, installation runs smoothly. `php-svg-lib` gets included too. `require_once ...` is not needed anymore.
@themao: I installed the bundle too, but (so far) can't see any real advantage. General question: What's the benefit of having a bundle over installing the (dompdf or any other) library directly?

**Comment by themao** [comment]
@ThomasLandauer the advantage is that it's already embedded as a service there and you can just get() it from the container. And, sometimes you get a fancy symfony config, but for this bundle it/s not the case, it uses those inc files

**Comment by ThomasLandauer** [comment]
@themao: Thanks, but I still don't see it ;-)
With the following 2 lines I can use dompdf anywhere in Symfony (can't see any advantage by calling it via the service container):

````php
use Dompdf\Dompdf;
$dompdf = new Dompdf();
````

And the options can now be set like this, which is (to me) more convenient than an .inc file:

````php
use Dompdf\Options;
$dompdfOptions = new Options();
$dompdfOptions->set('defaultFont', 'Courier');
$dompdf->setOptions($dompdfOptions);
````

However, I'm still not 100% convinced that I'm not missing **the** killer feature of the bundle ;-)

**Comment by themao** [comment]
@ThomasLandauer if you need the Dompdf more than once during runtime, create new instances with 'new Dompdf()' will lead to more intensive memory usage. Service in Symfony gets instantiated once and then gets reusable in any number of controllers/services/commands. 

In general you need a bundle when you need to configure it flexibly - let's say if it's using some html views that you need to override. In this particular case with Dompdf it's comfortable enough to call 'new Dompdf()', but looking on another more complicated bundles will show you the difference.

**Comment by bsweeney** [comment]
So I think running `composer require dompdf/dompdf:0.7.x@beta` is actually the correct command. It could be that you were only getting an earlier beta because of cached data or an existing install that met the requirement.

`dompdf/dompdf:v0.7.0-beta3` works because it is a reference to a specific release. If you have you [minimum stability](https://getcomposer.org/doc/04-schema.md#minimum-stability) set low enough (let's just say "beta") then you could also use `dompdf/dompdf:~0.7` to get the latest beta.

In the future you might try (in no particular order):

 - Clearing the Composer cache: `composer clear-cache`
 - Removing any existing install
 - Updating composer: `composer  self-update` then `composer update`

**Comment by raisen** [comment]
Strangely I'm getting the same error with CakePHP. I ran `composer require dompdf/dompdf:0.7.x@beta` to install dompdf. As @ThomasLandauer pointed out, it can be fixed by adding:

`require_once __DIR__ . '/../../../vendor/dompdf/dompdf/lib/Cpdf.php';`

I also tried the steps on https://github.com/dompdf/dompdf/issues/1132#issuecomment-201696037 but with no success.

**Comment by bsweeney** [comment]
@raisen I'm at a bit of a loss then. I spent some time reading through the Composer docs and testing myself and didn't have any issues. I wonder if it's possible something's interfering with the Composer autoloader.

We should be releasing 0.7.0 (stable) soon. I'll try to remember to post a note here so you can try the simpler `composer require dompdf/dompdf`.

**Comment by bsweeney** [comment]
The new release is now up. Hopefully the composer installation issues will be resolved

**Comment by blenderdeluxe** [comment]
> So it appears that this was solved already a year ago #919, but in 0.7.x@beta it definitely does not work. Googling around a little bit said the problem is that composer does not include files without namespaces (according to PSR-4) so I had to include it manually:
> 
> `"autoload": { "classmap": ["vendor/dompdf/dompdf/lib"]

> So it appears that this was solved already a year ago #919, but in 0.7.x@beta it definitely does not work. Googling around a little bit said the problem is that composer does not include files without namespaces (according to PSR-4) so I had to include it manually:
> 
> `"autoload": { "classmap": ["vendor/dompdf/dompdf/lib"] }`

Thanks! this work perfectly after 2 hours I update from "barryvdh/laravel-dompdf": "0.6.*", to "barryvdh/laravel-dompdf": "0.8.3", becuase the project is Laravel 5.2 and after composer update  and  "autoload": {
        "classmap": [
            "database",
            "vendor/dompdf/dompdf/lib"
        ],
        "psr-4": {
            "App\\": "app/"
        }
    },

and WORKS so fine!!!! thanks you save my day!
